### PR TITLE
Fix AddMerge and Recurrent layer memory issue.

### PR DESCRIPTION
**Pull request by zoq**
Fix memory issues as pointed out in https://github.com/mlpack/mlpack/issues/1187.

### CODE REVIEW

File: src/mlpack/methods/ann/layer/add_merge.hpp [hunk 1]
```diff
@@ -107,15 +114,33 @@ class AddMerge
...
+  //! Return the model modules.
+  // std::vector<LayerTypes<CustomLayers...> >& Model()
+  // {
+  //   if (model)
+  //   {
+  //     return network;
+  //   }
+
+  //   return empty;
+  // }
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/add_merge.hpp]: I think this commented code could be removed, but I'm not sure.  I guess maybe the layer should have a `Model()` function?

File: src/mlpack/methods/ann/layer/add_merge.hpp [hunk 2]
```diff
@@ -107,15 +114,33 @@ class AddMerge
...
+
   /**
    * Serialize the layer.
    */
   template<typename Archive>
   void serialize(Archive& ar, const unsigned int /* version */);
 
  private:
+  //! Parameter which indicates if the modules should be exposed.
+  bool model;
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/add_merge.hpp]: Hm, I guess I am a little confused about `model`---if it is `true`, then we delete the layers in the destructor.  Maybe a better name is `ownsLayers` or something?  I am not sure if I have understood fully so correct me if I'm wrong.

File: src/mlpack/methods/ann/rnn.hpp [hunk 3]
```diff
@@ -237,26 +237,18 @@ class RNN
...
-  template<typename LayerType>
-  void Add(const LayerType& layer) { network.push_back(new LayerType(layer)); }
-
-  /**
-   * Add a new layer to the model.
+  /*
+   * Add a new module to the model.
    *
-   * @param args The parameters to be passed to the layer constructor.
+   * @param layer The Layer to be added to the model.
```
Comment (rcurtin) [inline src/mlpack/methods/ann/rnn.hpp]: Oops, it looks like the changes make the documentation backwards---I guess this method should have `@param args` and the other should have `@param layer`.

File: src/mlpack/methods/ann/layer/add_merge.hpp (hunk 2 above)
Comment (zoq) [reply src/mlpack/methods/ann/layer/add_merge.hpp]: Good point, changed to ``ownsLayers``.

**Pull request merged**